source/Core/Src/Settings.cpp

```diff
@@ -12,85 +12,152 @@
+    {0, MAX_POWER_LIMIT, POWER_LIMIT_STEPS, POWER_LIMIT},           // PowerLimit
+    {0, 2, 1, REVERSE_BUTTON_TEMP_CHANGE},                          // ReverseButtonTempChangeEnabled
+    {5, TEMP_CHANGE_LONG_STEP_MAX, 5, TEMP_CHANGE_LONG_STEP},       // TempChangeLongStep
+    {1, TEMP_CHANGE_SHORT_STEP_MAX, 1, TEMP_CHANGE_SHORT_STEP},     // TempChangeShortStep
+    {0, 4, 1, 1},                                                   // HallEffectSensitivity
+    {0, 10, 1, 0},                                                  // AccelMissingWarningCounter
+    {0, 10, 1, 0},                                                  // PDMissingWarningCounter
+    {0, 0xFFFF, 0, 41431 /*EN*/},                                   // UILanguage
+    {0, 51, 1, 0},                                                  // PDNegTimeout
 
-bool restoreSettings() {
+};
+
+void saveSettings() { flash_save_buffer((uint8_t *)&systemSettings, sizeof(systemSettingsType)); }
+
+bool loadSettings() {
   // We read the flash
   flash_read_buffer((uint8_t *)&systemSettings, sizeof(systemSettingsType));
+  // Then ensure all values are valid
+  return sanitiseSettings();
+}
 
-  // if the version is correct were done
-  // if not we reset and save
-  if (systemSettings.version != SETTINGSVERSION) {
-    // probably not setup
-    resetSettings();
-    return true;
+bool sanitiseSettings() {
+  // For all settings, need to ensure settings are in a valid range
+  // First for any not know about due to array growth, reset them and update the length value
+  bool dirty = false;
+  if (systemSettings.padding != 0xFFFFFFFF) {
+    systemSettings.padding = 0xFFFFFFFF; // Force padding to 0xFFFFFFFF so that rolling forwards / back should be easier
+    dirty                  = true;
   }
-  return false;
-}
-// Lookup function for cutoff setting -> X10 voltage
-/*
- * 0=DC
- * 1=3S
- * 2=4S
- * 3=5S
- * 4=6S
- */
-uint8_t lookupVoltageLevel() {
-  if (systemSettings.minDCVoltageCells == 0)
-    return 90; // 9V since iron does not function effectively below this
-  else
-    return (systemSettings.minDCVoltageCells * systemSettings.minVoltageCells) + (systemSettings.minVoltageCells * 2);
+  if (systemSettings.length < (int)SettingsOptions::SettingsOptionsLength) {
+    dirty = true;
+    for (int i = systemSettings.length; i < (int)SettingsOptions::SettingsOptionsLength; i++) {
+      systemSettings.settingsValues[i] = 0xFFFF; // Ensure its as if it was erased
+    }
+    systemSettings.length = (int)SettingsOptions::SettingsOptionsLength;
+  }
+  for (int i = 0; i < (int)SettingsOptions::SettingsOptionsLength; i++) {
+    // Check min max for all settings, if outside the range, move to default
+    if (systemSettings.settingsValues[i] < settingsConstants[i].min || systemSettings.settingsValues[i] > settingsConstants[i].max) {
+      systemSettings.settingsValues[i] = settingsConstants[i].defaultValue;
+
+      dirty = true;
+    }
+  }
```

Comment:
Am I right that this might not reset all configurations when upgrading from 2.15 or previous versions? It doesn't explicitly check for `<= SETTINGSVERSION`. It seems possible that some of the old setting data may have bit patterns that can sneak past the sanitizing code.

Reply:
@Ralim
Good evening Sir,
maybe you've missed this one, so I am humbly asking you for an answer. 😅

Reply:
I did see this; but did do a bit of messing around.
@alvinhochun you are right that its possible for some things to fall through if their values line up nicely. Though in my testing I didnt spot any that did 😓 And felt easier to not bother with keeping it for more checks... Might be crazy though on that :)

Reply:
When I flashed the new firmware, I did notice that the advanced standby screen was enabled, so that might be one that can get through.